### PR TITLE
Add undo feature in Editing plugin

### PR DESCRIPTION
**Pull request by arnaud-morvan**

### CODE REVIEW

File: geoext.ux/ux/FeatureEditing/lib/GeoExt.ux/FeatureEditorGrid.js [hunk 1]
```diff
@@ -238,7 +238,7 @@ GeoExt.ux.FeatureEditorGrid = Ext.extend(Ext.grid.EditorGridPanel, {
                 this.saveButton
             ]
         });
-        this.dirty = feature.state == OpenLayers.State.INSERT;
+        this.dirty = !!feature.state;
```
Comment (sbrunner) [inline geoext.ux/ux/FeatureEditing/lib/GeoExt.ux/FeatureEditorGrid.js]: This test looks relay strange for me, Why do you change it ?

File: geoext.ux/ux/FeatureEditing/lib/GeoExt.ux/FeatureEditorGrid.js [hunk 2]
```diff
@@ -309,6 +310,28 @@ GeoExt.ux.FeatureEditorGrid = Ext.extend(Ext.grid.EditorGridPanel, {
                 return false;
             }
         }, this);
+
+        this.undoStack = [];
+        if (!!feature) {
```
Comment (sbrunner) [inline geoext.ux/ux/FeatureEditing/lib/GeoExt.ux/FeatureEditorGrid.js]: How can this test be false ?

File: geoext.ux/ux/FeatureEditing/lib/GeoExt.ux/FeatureEditorGrid.js [hunk 3]
```diff
@@ -318,11 +341,77 @@ GeoExt.ux.FeatureEditorGrid = Ext.extend(Ext.grid.EditorGridPanel, {
...
     onFeaturemodified: function(e) {
         if(e.feature === this.store.feature) {
-            this.dirty = true;
-            this.saveButton.setDisabled();
+            if (e.action == 'undo') {
+                // Do nothing
+            }
+            else {
+                // Don't stash twice on first modification
+                if (!!e.feature.state) {
```
Comment (sbrunner) [inline geoext.ux/ux/FeatureEditing/lib/GeoExt.ux/FeatureEditorGrid.js]: Also looks strange ...

File: geoext.ux/ux/FeatureEditing/lib/GeoExt.ux/FeatureEditorGrid.js [hunk 4]
```diff
@@ -318,11 +341,77 @@ GeoExt.ux.FeatureEditorGrid = Ext.extend(Ext.grid.EditorGridPanel, {
...
         }
     },
 
+    /** private: method[stash]
+     *  :param feature: ``OpenLayers.Feature`` The feature to stash.
+     *
+     *  Stash feature state and properties for the undo functionality.
+     */
+    stash: function(feature) {
+        if (!!this.previous) {
```
Comment (sbrunner) [inline geoext.ux/ux/FeatureEditing/lib/GeoExt.ux/FeatureEditorGrid.js]: `this.previous !== null` is relay better!, and it should be initialized to to null and documented

**Comment by sbrunner** [comment]
I stop my review because I see many boolean operation on non boolean type, this should be avoid ...

**Comment by arnaud-morvan** [comment]
I've removed all boolean tests on feature.state, replaced by explicit tests.
I've initialized and documented this.previous and this.undoStack properties.

**Comment by sbrunner** [comment]
Excellent, for me it's OK to merge :-)

**Pull request merged**